### PR TITLE
Add client-side session id generation and new user activities

### PR DESCRIPTION
**Pull request by huffman**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8b3b62aad7ac3861f044fdaebf8ecdc5f6388de5 available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 56c84bbc27cf34cf1a173d67b28e1c0fdc8cc871 available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bce05df56b5718cc1f21a1238d8c77effe667800 available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c2ebcd1f77234350a974a03b30c1aac1223e7f8a available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by huffman** [comment]
Test plan coming shortly...

**Comment by huffman** [comment]
We are adding new usage data so we can find out where people are having a poor experience in High Fidelity. To start, make sure you are sending usage data by going to Settings > General, then under the "Privacy" section, make sure "Send data" is checked. Once that is done, follow these steps:

1. Run this script: https://gist.githubusercontent.com/huffman/1aad59b6c1b43ffdd666a8da095f989e/raw/aa84b84718705995d84e3a95c12a86acd3e717ec/nullScript.js
2. Open your address bar and go to "playa"
3. Go to an area where other avatar(s) are. You can use the friends list to do this
4. Enable edit (you don't need to do anything in it)
5. Open the examples window (you don't need to download anything)
6. Go to your local server and try to get the audio server to mute you. You can usually get it to mute you by blowing into your mic for an extended period of time, but it can sometimes be tricky.

Once you have completed these steps, let me know, and I can verify things on my end.

**Comment by kevinmthomas-carpool** [comment]
testing

**Comment by kevinmthomas-carpool** [comment]
## QA Test
**QA Tester**: KevinMThomas
**QA Pass**: **QA Pass** FAILURE
**Additional Test Results + Observations**: Ran ```https://gist.githubusercontent.com/huffman/1aad59b6c1b43ffdd666a8da095f989e/raw/aa84b84718705995d84e3a95c12a86acd3e717ec/nullScript.js``` and opened address bar and went to ```playa```.  Crashed.  Relogged and re-ran script.  Re-visited ```playa```.  Successfully able to edit examples window and saw web window appear with items.  Went back to ```localhost```.  Was unable to get mic to mute by blowing air into mic over a 30 second window with short breaks to breathe.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 154ccb8932095f2f180cda149eb912c68fef0348 available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0b311568c8e441b55eb0f147322e009bf3404708 available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a77dea904876725d583638e8d382ff14917c3aef available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6d753e317b60dfb3ee9b6c5b3be826b11e999bbe available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

File: libraries/shared/src/SharedUtil.cpp [hunk 1]
```diff
@@ -843,3 +844,27 @@ void printSystemInformation() {
...
+
+    PROCESS_MEMORY_COUNTERS_EX pmc;
+    if (!GetProcessMemoryInfo(GetCurrentProcess(), reinterpret_cast<PROCESS_MEMORY_COUNTERS*>(&pmc), sizeof(pmc))) {
+        return false;
+    }
+    info.processUsedMemoryBytes = pmc.PrivateUsage;
+    info.processPeakUsedMemoryBytes = pmc.PeakPagefileUsage;
+#endif
+
+    return true;
```
Comment (ZappoMan) [inline libraries/shared/src/SharedUtil.cpp]: should we return false for non-windows OS?

Comment (huffman) [inline libraries/shared/src/SharedUtil.cpp]: Good catch, fixed

**Comment by huffman** [comment]
Making one more change to the sixense plugin. Right now it is marked as supported if you don't have the base plugged in.

File: libraries/networking/src/UserActivityLogger.cpp [hunk 2]
```diff
@@ -125,6 +125,21 @@ void UserActivityLogger::changedDomain(QString domainURL) {
 }
 
 void UserActivityLogger::connectedDevice(QString typeOfDevice, QString deviceName) {
+    static QStringList DEVICE_BLACKLIST = {
```
Comment (ZappoMan) [inline libraries/networking/src/UserActivityLogger.cpp]: I don't entirely understand why you're black listing all of these... like SDL, Neuron... don't we want to know that they are available?

Comment (huffman) [inline libraries/networking/src/UserActivityLogger.cpp]: From the perspective of `isSupported()` they are always available. I'm looking to see if there is a better way to handle these particular cases. For SDL2, for instance, we could track each individual connected device in `init()` - the device name in this case would not be "SDL2."

Comment (ZappoMan) [inline libraries/networking/src/UserActivityLogger.cpp]: ok

Comment (huffman) [inline libraries/networking/src/UserActivityLogger.cpp]: Ok, each plugin has it's own implementation, and not all know if the physical device is connected at `isSupported()`, so I had to change the way device connections are detected. I added a `deviceConnected` signal to `Plugin`, each plugin needs to be implemented separately to emit this signal when the device is known to be connected.

**Comment by huffman** [comment]
This is ready to be tested again once it builds.

**Comment by ZappoMan** [comment]
Let's make sure we test this on the Mac too...

**Comment by hifi-gustavo** [comment]
Build console output for commit: 70be7b3273b51fcbca4e3772b48f33add7317254 available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by huffman** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4e52521f8cd05d4a30a0cc219c119d9f1fc37d79 available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 44aab9f255edc18d3f05de43ff5cd3bccd116ff3 available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a7f30ced29516864b4c8f94367c1b02c3e23e534 available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by kevinmthomas-carpool** [comment]
## QA Re-Test
**QA Tester**: KevinMThomas
**QA Pass**: **QA Pass** FAILURE
**Additional Test Results + Observations**: Ran ```https://gist.githubusercontent.com/huffman/1aad59b6c1b43ffdd666a8da095f989e/raw/aa84b84718705995d84e3a95c12a86acd3e717ec/nullScript.js``` and opened address bar and went to ```playa```.    Successfully able to edit examples window and saw web window appear with items.  The first test a few days ago I crashed where in this instance I did not. Visited ```localhost```.  Was unable to get mic to mute by blowing air into mic over a long period with short breaks to breathe.  Additionally attempted tapping hard on mic to try to generate a spike in volume.  Unable to get it to mute.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2621add8e30b8a9d766c74f5eabdfe92acc14d64 available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by huffman** [comment]
test this please

**Comment by huffman** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2621add8e30b8a9d766c74f5eabdfe92acc14d64 available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by huffman** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2621add8e30b8a9d766c74f5eabdfe92acc14d64 available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2621add8e30b8a9d766c74f5eabdfe92acc14d64 available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by kevinmthomas-carpool** [comment]
## QA Re-Test
**QA Tester**: KevinMThomas
**QA Pass**: **QA Pass** FAILURE
**Additional Test Results + Observations**: Ensured ```send data``` was enabled in ```General```.  Ran script: ```https://gist.githubusercontent.com/huffman/1aad59b6c1b43ffdd666a8da095f989e/raw/aa84b84718705995d84e3a95c12a86acd3e717ec/nullScript.js?v2``` and opened my  address bar went to```playa```.  Enabled edit, opened the examples window.  Went to ```localhost``` attempted to get the audio server to mute unsuccessfully. Blew air for an extended period of time in both desktop mode and then in Oculus rift mode.  I then disabled the noise gate by ```Developer > Audio > Audio Noise Reduction```  A Sixense, Oculus were plugged in an active.   Pressed escape to go away, then any key to go back to active.  I re-ran through the above steps and was unable to get it to mute.  Even attempted screaming.  Saw peak at several seconds but a mute did not occur.

**Comment by huffman** [comment]
Ok, thanks @kevinmthomas-carpool . If the only thing that failed was the muting then we can consider this a pass. I've been able to do it locally, and the actual server-muting behavior is not what was implemented here.

**Comment by huffman** [comment]
I tested successfully on OSX as well

**Comment by kevinmthomas-carpool** [comment]
@huffman Yes to confirm everything else works.

**Comment by ZappoMan** [comment]
Tentative - Dev Approve - waiting on @huffman to add the support for the OculusLegacy plugin for mac DK2 support.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2233b5ba64f7c2339a802ccbb95dcc1fd831096c available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9ff742c97c4c400a0eea97ed366389d8005843c9 available at: https://gist.github.com/642c745f9c3962e3f06286fe06a648af<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8084.dmg

**Comment by ZappoMan** [comment]
tested on mac

**Pull request merged**